text/1615-cargo-platform-standards.md

```diff
@@ -0,0 +1,170 @@
+   them.
+
+This makes Cargo use platform-specific directories for new installs while
+retaining compatibility for the old directory layout. It also allows one to
+keep all Cargo related data in one place if one wishes to.
+
+## Windows
+
+We'll obtain each of the following directories using the correct API.
+
+```
+cache:    AppData\Local\Temp\Cargo
+config:   AppData\Roaming\Cargo
+binaries: AppData\Local\Programs\Cargo
+```
+
+## Unixy systems (macOS, Linux, BSDs)
+
+Here, we're following the [XDG specification](https://specifications.freedesktop.org/basedir-spec/basedir-spec-0.7.html).
+By default, if no further variables are set, this means that we'll be using the
+following subdirectories of the home directory.
+
+```
+cache:    .cache/cargo
+config    .config/cargo
+binaries: .local/bin
+```
+
+**There is currently an on-going discussion about standardizing the location of
+`.local/bin` together with a new XDG variable `XDG_BIN_HOME`. The
+implementation of this RFC should be delayed until that discussion has finished
+and use the result. This RFC will be amended with that result.**
+
+
+## Rustup
+
+Rustup will replicate Cargo's priorisation algorithm. If the results differ
+from what the executed version of Cargo will do, Rustup will add environment
+variables `CARGO_BIN_DIR`, `CARGO_CACHE_DIR`, `CARGO_CONFIG_DIR` for the new
+versions of Cargo, and add symlinks for the old versions of Cargo.
+
+
+## New subcommand
+
+Cargo (and Rustup) are going to gain a new subcommand, `cargo dirs`. It will
+display the directories currently in use, in a human-readable format. In order
+to support other programs and scripts, this subcommand will also have switches
+to print the data in machine-readable form, at least `--json` for JSON output
+and maybe `--shell` for env-compatible output.
+
+Example JSON (specifics left to the implementation):
+```
+{
+  "bin_dir": "C:\\Users\\User\\AppData\\Local\\Programs\\Cargo",
+  "cache_dir": "C:\\Users\\User\\AppData\\Local\\Temp\\Cargo",
+  "config_dir": "C:\\Users\\User\\AppData\\Roaming\\Cargo"
+}
+```
+
+Example (env-compatible):
+```
+CARGO_BIN_DIR=/home/user/.local/bin
+CARGO_CACHE_DIR=/home/user/.cache/cargo
+CARGO_CONFIG_DIR=/home/user/.config/cargo
+```
+
+
+# Drawbacks
+
+* This increases the complexity of where to find the files Cargo uses. This can
+  be alleviated by providing library functions for Cargo related tools and a
+  command line to tell which paths are in use.
+
+* It can also complicate tutorials that tell users how Cargo works. The
+  tutorials should probably reference the Cargo subcommand that displays the
+  used locations. However it's still more complicated than static directory
+  name (it's in a weird location for Windows users though).
+
+
+# Alternatives
+
+* macOS could also use the `Library` folder for storing its data. This is mostly
+  done by UI applications.
+
+* One could only change the Windows paths, as the Windows integration is
+  currently the worst.
```

Comment:
The most notable thing about the Windows integration is probably the non-hiddenness of the folders, I should have that fixed in a release or two at most now it has been pointed out.